### PR TITLE
scripts: qemu-check.sh: Refactor and improve script 

### PR DESCRIPTION
**Pull request by crazoes**

### CODE REVIEW

**Comment by crazoes** [comment]
Furthers https://github.com/thesofproject/sof/issues/2631

**Comment by crazoes** [comment]
Build is succeeding for `imx8`, `imx8x`  and failing for `imx8m`. 

```
qemu-system-xtensa: terminating on signal 15 from pid 56 (timeout)
```

I can see qemu is timing out and a `dump-imx8m.txt` file is tailed in the logs. However, I am not sure what to make of that information. Any suggestions on how to debug this?

**Comment by dbaluta** [comment]
@crazoes is the build working without your patches?

**Comment by crazoes** [comment]
> @crazoes is the build working without your patches?

Didn't check it before, assumed builds are always green on master. But the same issue seems to be there on master, https://travis-ci.org/github/thesofproject/sof/jobs/676116208. So probably not a bug in my patches.

**Comment by dbaluta** [comment]
@crazoes indeed the bug is from some other part. @dianacretu can you have a look? Looks like imx8m no longer boots with qemu.

@crazoes we are mostly off until Wednesday due to orthodox Easter. Will try to have a look as soon as possible.

File: scripts/qemu-check.sh [hunk 1]
```diff
@@ -7,110 +7,133 @@ READY_MSG="6c 00 00 00 00 00 00 70"
...
 else
-	for args in $@
-	do
-		for i in ${SUPPORTED_PLATFORMS[@]}
-		do
-			if [ $i == $args ]
-			then
-				PLATFORMS+=$i" "
+	for arg in "$@"; do
+		platform=none
```
Comment (marc-hb) [inline scripts/qemu-check.sh:30]: Can you change `none` to `unknown`? This threw me off.

File: scripts/qemu-check.sh [hunk 2]
```diff
@@ -7,110 +7,133 @@ READY_MSG="6c 00 00 00 00 00 00 70"
...
+				PLATFORMS=("${PLATFORMS[@]}" "$i")
+				platform=$i
+				shift
+				break
 			fi
 		done
+		if [ "$platform" = "none" ]; then
+			echo "Error: Unknown platform specified: $arg"
+			echo "Supported platforms: ${SUPPORTED_PLATFORMS[*]}"
+			exit 1
```
Comment (marc-hb) [inline scripts/qemu-check.sh:42]: When there is more than one `echo ERRROR; exit 1` you should (re-)use a `die()` function like this one:
```shell
die()
{
    >&2 printf '%s: ERROR: ' "$0"
    >&2 printf "$@"
    exit 1
}
```
Then you can shorten the input validation checks into one-liners like this:
```shell
[ ${#PLATFORMS[@]} -gt 0 ] || die  ''No platforms specified. Supported are: %s\n'\
		"${SUPPORTED_PLATFORMS[*]}"
```

File: scripts/qemu-check.sh [hunk 3]
```diff
@@ -7,110 +7,133 @@ READY_MSG="6c 00 00 00 00 00 00 70"
...
+		fi
 	done
 fi
 
 # check target platform(s) have been passed in
 if [ ${#PLATFORMS[@]} -eq 0 ];
 then
-	echo "Error: No platforms specified. Supported are: ${SUPPORTED_PLATFORMS[@]}"
+	echo "Error: No platforms specified. Supported are: "\
```
Comment (marc-hb) [inline scripts/qemu-check.sh:50]: Did you try to trigger and test this (new) error message? I don't think you can. If this error cannot be triggered then it's pointless, remove it.

File: scripts/qemu-check.sh [hunk 4]
```diff
@@ -7,110 +7,133 @@ READY_MSG="6c 00 00 00 00 00 00 70"
...
-	for args in $@
-	do
-		for i in ${SUPPORTED_PLATFORMS[@]}
-		do
-			if [ $i == $args ]
-			then
-				PLATFORMS+=$i" "
+	for arg in "$@"; do
+		platform=none
+		for i in "${SUPPORTED_PLATFORMS[@]}"; do
```
Comment (marc-hb) [inline scripts/qemu-check.sh:31]: Nit: `i` -> `sp`

File: scripts/qemu-check.sh (hunk 3 above)
Comment (crazoes) [reply scripts/qemu-check.sh:50]: Right, because if nothing is specified, it uses all platforms..

File: scripts/qemu-check.sh (hunk 2 above)
Comment (crazoes) [reply scripts/qemu-check.sh:42]: After i remove the exit for no platforms which is unreachable, this would be the only one

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
@crazoes thanks, this is all good stuff. If you have time, are you interested in upstreaming the qemu DSP support in qemu master ? The code is mostly ready, but it will need checked/scanned/fixed and then sent upstream to qemu. There will be some code changes involved so you would be able to add yourself to the authors for the ADPS qemu code.

Repo is here https://github.com/thesofproject/qemu

Code is in the hw/adsp/dsp directory and target/xtensa/ directories. Please send me a email if interested. Thanks.

**Comment by crazoes** [comment]
> @crazoes thanks, this is all good stuff. If you have time, are you interested in upstreaming the qemu DSP support in qemu master ? The code is mostly ready, but it will need checked/scanned/fixed and then sent upstream to qemu. There will be some code changes involved so you would be able to add yourself to the authors for the ADPS qemu code.
> 
> Repo is here https://github.com/thesofproject/qemu
> 
> Code is in the hw/adsp/dsp directory and target/xtensa/ directories. Please send me a email if interested. Thanks.

I don't really understand what this work would entail other than the part that we need to get something from sof's qemu fork merged into the main qemu repository. Looks like a good opportunity to get more familiar with sof outside the script contributions I have done so far.

**Comment by lgirdwood** [comment]
> I don't really understand what this work would entail other than the part that we need to get something from sof's qemu fork merged into the main qemu repository. Looks like a good opportunity to get more familiar with sof outside the script contributions I have done so far.

Yes, this is exactly what is needed. The code also needs a review for obvious bugs like you have been doing with SOF and some refactoring/cleanup to de-duplicate/optimise and align with qemu master APIs.

 I would probably start with the baytrail target in target/xtensa/core-baytrail* (as this is simplest) and the add the baytrail files from hw/adsp/dsp.

**Comment by dbaluta** [comment]
@crazoes indeed the build on imx8m still fails even https://github.com/thesofproject/qemu/pull/38 was merged.

So, as you said I think debugging this will help you better understand SOF. So, as Liam said
you can get involved into upstreaming Qemu changes.

As for i.MX8M change you can start by:

1) get SOF repo
2) get qemu repo  https://github.com/thesofproject/qemu/
3) build qemu as described in Chapter 'Optional. Build the Qemu DSP emulator from source' from here https://www.alsa-project.org/wiki/Firmware#Installing_the_SDK
4) Test imx8m boot with the following command

```
./xtensa-softmmu/qemu-system-xtensa -cpu imx8m -M adsp_imx8m -nographic -kernel ../sof.git/build_imx8m_gcc/src/arch/xtensa/sof-imx8m.ri
```
5) I think there is a regression in the Qemu or SOF, so try going back few patches for both SOF/Qemu repos.

Cc: @dianacretu

**Comment by marc-hb** [comment]
> I think there is a regression in the Qemu or SOF, so try going back few patches for both SOF/Qemu repos.

If you look at the `sof/.travis.yml` config https://travis-ci.org/github/thesofproject/sof/builds/676971053/config you can see this Travis doesn't do any continuous integration of QEmu, it merely uses the QEmu version provided in the docker image.

How are changes to the Docker image tracked and where?

**Comment by dbaluta** [comment]
> How are changes to the Docker image tracked and where?

paging @xiulipan

**Comment by marc-hb** [comment]
OK, there are much more difficult detective jobs: https://hub.docker.com/r/thesofproject/sofqemu/tags

There was indeed an update by @xiulipan 5 days ago.

**Comment by crazoes** [comment]
> > I think there is a regression in the Qemu or SOF, so try going back few patches for both SOF/Qemu repos.
> 
> If you look at the `sof/.travis.yml` config https://travis-ci.org/github/thesofproject/sof/builds/676971053/config you can see this Travis doesn't do any continuous integration of QEmu, it merely uses the QEmu version provided in the docker image.
> 
> How are changes to the Docker image tracked and where?

Maybe this is the reason:

> Building the container from DockerFile will take more than 2 hours, so we recommend using the pre-built image.

^^ Note from docs at https://thesofproject.github.io/latest/getting_started/build-guide/build-with-docker.html

I suppose we could at best try to be continously integrating with latest qemu image which gets built and pushed to docker hub by master builds on travis on qemu repo. Although, I am not sure if travis allows jobs that take this long without it killing the jobs.

**Comment by xiulipan** [comment]
@crazoes @dbaluta  @marc-hb I tried to enable the Dockerhub auto build, but it seems could not satisfy our need. So the Docker image update is all handled manually.
I just updated the latest Docker Image. Will try to enable a CI/CD for the QEMU and Build Docker image when I get some time.

Please ping me when there is any Docker image related issue.

**Comment by crazoes** [comment]
@xiulipan can you explain why dockerhub auto build is not fulfilling our needs? Maybe we can help solve the issue

**Comment by marc-hb** [comment]
Guys, we're digressing from a digression. One advantage thing about a static Docker image is fewer moving parts less frequently, which means easier to root cause why the `imx8m` test fails right now _which turns Travis red in every single PR_ (including this one). That's the top priority right now, we can discuss Docker version control somewhere else another time.

@xiulipan , did your most recent Docker upgrade broke the `imx8m` test? If yes please revert it now.

**Comment by xiulipan** [comment]
@marc-hb Yes, I found there is some regression on imx8m. But there is already some fix for it. So I just update the Docker image to the the newest version.

**Comment by marc-hb** [comment]
`imx8m` test passing again in most recent PR #2834, thanks @xiulipan !

@dbaluta can you double check the logs to make sure there's no new warning or anything?

**Comment by dbaluta** [comment]
@marc-hb looks good now. Thanks everybody for cooperation!

**Pull request merged**